### PR TITLE
Fix insideCallback inconsistency in NetworkPeer

### PR DESCRIPTION
**Pull request by quantumagi**
Comparing this code:

`ProcessMessageAsync`:
```
            try
            {
                this.onDisconnectedAsyncContext.Value = new DisconnectedExecutionAsyncContext();

                await this.MessageReceived.ExecuteCallbacksAsync(this, message).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                this.logger.LogCritical("Exception occurred while calling message received callbacks: {0}", e.ToString());
                this.logger.LogTrace("(-)[EXCEPTION_CALLBACKS]");
                throw;
            }
            finally
            {
                if (this.onDisconnectedAsyncContext.Value.DisconnectCallbackRequested)
                    this.onDisconnected(this);

                this.onDisconnectedAsyncContext.Value = null;
            }
```
and `OnStateChangedAsync`:
```
        private async Task OnStateChangedAsync(NetworkPeerState previous)
        {
            bool insideCallback = this.onDisconnectedAsyncContext.Value == null; // Should be !=
            if (!insideCallback)
                this.onDisconnectedAsyncContext.Value = new DisconnectedExecutionAsyncContext();

            try
            {
                await this.StateChanged.ExecuteCallbacksAsync(this, previous).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                this.logger.LogError("Exception occurred while calling state changed callbacks: {0}", e.ToString());
                throw;
            }
            finally
            {
                if (!insideCallback)
                {
                    if (this.onDisconnectedAsyncContext.Value.DisconnectCallbackRequested)
                        this.onDisconnected(this);

                    this.onDisconnectedAsyncContext.Value = null;
                }
            }
        }
```

we notice that `insideCallback` is only used in the one case. There appears to be an assumption that `ProcessMessageAsync` can never be called while another `ProcessMessageAsync` or `OnStateChangedAsync` is being processed. Either that or `AsyncLocal` is coming to the rescue. The code will be much less opaque and easier to verify if we remove the need for `AsyncLocal`. Besides `AsyncLocal` can't prevent disconnect from being called twice for two simultaneous `ProcessMessageAsync` calls happening on separate threads. Assuming that won't ever happen it would still be better to have thread-safe code rather than relying on assumptions about usage.

Note that `!insideCallback` can be read as `iCreatedTheContext`. Based on that observation the apparent purpose of `insideCallback` is to ensure that only the caller that first creates the context is responsible for processing disconnections and removing the context. This should be equally applicable in both cases, so this PR fixes `ProcessMessageAsync` to promote such consistency. 

We will probably have to remove `AsyncLocal` as part of this PR.

### CODE REVIEW

**Comment by quantumagi** [comment]
Merge this first: https://github.com/stratisproject/StratisBitcoinFullNode/pull/3859

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs [hunk 1]
```diff
@@ -530,10 +530,12 @@ private async Task ProcessMessageAsync(IncomingMessage message, CancellationToke
                 return;
             }
 
-            try
-            {
+            bool insideCallback = this.onDisconnectedAsyncContext.Value != null;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: I believe processes message is always processed by a single thread (sequentially) and will never be called from one of the behaviours so `this.onDisconnectedAsyncContext.Value` will never have a value (if I am not mistaken)

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs [hunk 2]
```diff
@@ -544,10 +546,13 @@ private async Task ProcessMessageAsync(IncomingMessage message, CancellationToke
...
             {
-                if (this.onDisconnectedAsyncContext.Value.DisconnectCallbackRequested)
-                    this.onDisconnected(this);
+                if (!insideCallback)
+                {
+                    if (this.onDisconnectedAsyncContext.Value.DisconnectCallbackRequested)
+                        this.onDisconnected(this);
 
-                this.onDisconnectedAsyncContext.Value = null;
+                    this.onDisconnectedAsyncContext.Value = null;
```
Comment (dangershony) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: This will reset the AsuncLocal for the next message to be processed.

**Comment by dangershony** [comment]
>We will probably have to remove AsyncLocal as part of this PR. 

That may be true how ever in that case a design needs to be proposed, I agree AsyncLocal is not easy or intuitive to grasp.

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: Perhaps we could put this PR on ice and instead create a PR that adds a `Guard.Assert` to ensure that the value is always null.

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: See previous comment.

**Comment by quantumagi** [comment]
> Perhaps we could put this PR on ice and instead create a PR that adds a Guard.Assert to ensure that the value is always null. Alternatively we can throw an exception.

**Comment by dangershony** [comment]
The issue in general is that every behaviour can call back in to the NetworkPeer and call disconnect or change state so there is a recursive logic here.

IF a design is to be proposed to fix this then one will have to look at all the places where a behaviour can call in to the NetworkPeer and change that with delayed execution (i.e a behaviour can not change state or disconnect but signal those actions so only when the message received loop is finished it will act on that, or if an external event called Disconnect)

File: src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs [hunk 3]
```diff
@@ -522,10 +522,13 @@ private async Task ProcessMessageAsync(IncomingMessage message, CancellationToke
                 return;
             }
 
-            try
-            {
+            // Creates a context that can be used to postpone / flag disconnect.
+            bool iCreatedContext = this.onDisconnectedAsyncContext.Value != null;
```
Comment (zeptin) [inline src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs]: Shouldn't this be `==` to match the logic in `OnStateChangedAsync`?

**Pull request merged**